grammar.js

```diff
@@ -879,7 +879,23 @@ module.exports = grammar({
         $.identifier
       )),
       '!',
-      $.token_tree
+      alias($.delim_token_tree, $.token_tree)
+    ),
+
+    delim_token_tree: $ => choice(
+      seq('(', repeat($._delim_tokens), ')'),
+      seq('[', repeat($._delim_tokens), ']'),
+      seq('{', repeat($._delim_tokens), '}')
+    ),
+
+    _delim_tokens: $ => choice(
+      $._non_delim_token,
+      alias($.delim_token_tree, $.token_tree),
+    ),
+
+    _non_delim_token: $ => choice(
+      $._non_special_token,
+      '$'
```

Comment:
do we need other characters here? This is needed just for '$'?

Reply:
I believe `$` is the only character we need to add here. `_non_special_token` is analogous to the first branch of the [`MacroMatch` construction](https://doc.rust-lang.org/reference/macros-by-example.html#macros-by-example) in the reference, which allows any token other than `$` and delimiters.